SC-16400 add default ARs in integrations

cpipilas opened the pull request:
- This PR adds a section for default alert rules in all integrations supporting them.
- Each alert rule contains a description, an example and a list of actions to be taken. 
- Each action is limited to a short description on what the user should do.

cpipilas commented on docs/integration/opensearch-integration.md:
```diff
@@ -40,6 +40,135 @@ OpenSearch runs within a Java Virtual Machine (JVM) and [monitoring JVM](https:/
+
+This alert rule continuously monitors swap usage in an OpenSearch environment by tracking the rate of swap input/output operations. When swap usage exceeds the specified threshold, it triggers a warning (WARN priority). The minimum delay between consecutive alerts triggered by this alert rule is set to 10 minutes.
+
+Suppose the swap usage on a node in the OpenSearch cluster suddenly spikes, indicating potential resource constraints or performance issues. When this happens, the alert rule checks for swap usage over the last 10 minutes. The alert is triggered when the sum of swap input/output operations within the specified timeframe is more than 1 (the threshold value).
```
@radu-gheorghe I think I'm wrong here. Can you confirm the condition for triggering this alert?

radu-gheorghe replied:
Actually, I think any swap usage (even if you read 1 byte) would trigger the alert. The idea is that OpenSearch needs OS cache (read: free memory) to work with the index. If the OS swaps the JVM memory, it means one of the following: there's not enough RAM on the box or (more commonly) the OS decides to swap a chunk of RAM preemptively (because of an inappropriate `vm.swappiness` setting. Either way, it could introduce significant (read: unacceptable) delays when OpenSearch wants to access that piece of memory, so it's a no-no to swap.

cpipilas replied:
Thanks @radu-gheorghe! I've updated the AR.

radu-gheorghe commented on docs/integration/akka.md:
```diff
@@ -5,6 +5,45 @@ description: Sematext APM monitoring integration for Akka actors, dispatchers, a
+- Check the status of the network infrastructure to determine if there are any issues such as congestion, packet loss, or connectivity problems affecting data reception
+- Review application logs and system metrics for any errors or warnings that may be impacting network data reception
+
+### Network transmitted anomaly
```
This is fine, but maybe we can just say it's the same as ingress, but this is egress? :)

cpipilas replied:
Yep, I'm trying to give all the details for each alert rule. There is a bunch of repeated text, but it's like each piece of info stands alone and helps connect with SC later on.

radu-gheorghe replied:
Got it. Thanks!

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+#### Actions to take
+
+- Check the status of nodes within the Elasticsearch cluster for any nodes that may be offline, unavailable, or experiencing issues
```
You'll probably want to check the logs of the node that went down and/or logs of the master node.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+This alert rule continuously monitors the usage of Java's old generation heap memory in an Elasticsearch environment, triggering a warning if the usage exceeds 97%. The minimum delay between consecutive alerts triggered by this alert rule is set to 10 minutes.
+
+Suppose the Java old generation heap memory usage in the Elasticsearch environment starts increasing and eventually exceeds 97% over a 5-minute period.
```
I would add that this shouldn't happen in a healthy environment. It's likely that the node will either face an out of memory exception or run into the Parent Circuit Breaker. Either way, you'd have unexpected failures.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+#### Actions to take
+
+- Analyze the application's memory usage patterns for any memory leaks
+- Review and optimize the Java Virtual Machine (JVM) configuration, including heap size settings, garbage collection algorithms, and memory management parameters
```
Link to some of our blog posts on the topic?

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+- Analyze the application's memory usage patterns for any memory leaks
+- Review and optimize the Java Virtual Machine (JVM) configuration, including heap size settings, garbage collection algorithms, and memory management parameters
+- Monitor system resources, including memory, CPU, and disk I/O, and consider scaling them up if necessary
```
I'd remove this, we're already monitoring :)

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+This alert rule continuously monitors the field data size in an Elasticsearch cluster and triggers a warning if the field data size exceeds a certain threshold (20 in this case). The minimum delay between consecutive alerts triggered by this alert rule is set to 10 minutes.
+
+Suppose the field data size in the Elasticsearch cluster increases significantly due to increased query loads or inefficient queries. When this happens, the alert rule checks the field data size over the last 10 minutes. If it exceeds 20% of the JVM heap committed memory during this timeframe, the alert rule triggers a warning.
```
This is getting repetitive. In this particular case, I'd say that significant field data usage points to a misconfiguration. Normally, you'd only use field data for global ordinals. If you're using >20%, you probably do sorting/aggregations on a text field with field data enabled. Which is dangerous (you may run out of heap on an expensive query). So you'd want to pre-process the data in the pipeline before ES and do your sorting/aggregations on doc_values instead.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+#### Actions to take
+
+- Analyze the query patterns and usage that contribute to the increased field data size. Look for inefficient or resource-intensive queries that may need optimization
```
_cat/fielddata will tell you which fields use more field data.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+#### Actions to take
+
+- Analyze the query patterns and usage that contribute to the increased field data size. Look for inefficient or resource-intensive queries that may need optimization
+- Optimize queries to reduce the amount of field data loaded into memory. This may involve restructuring queries, using filters, or optimizing aggregations to minimize memory usage
```
You can't fix this by changing queries, see my earlier comment.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+- Consider scaling up the resources allocated to the Elasticsearch cluster, such as increasing the JVM heap size, to accommodate the increased field data size
+
+
+### Tripped parent circuit breaker
```
Parent Breaker has two modes:
- the default, on "real data". Meaning it looks at the current heap usage before doing an operation. If heap usage > threshold --> operation is not run, because you're in danger of running out of heap
- real_data=false --> just the sum of all the other breakers. I.e. if an operation tries to load memory that would use more than X% of heap across all breakers (e.g. field data, in-flight requests...), that operation isn't run

cpipilas replied:
Thanks Radu. So we alert on any Parent circuit breaker exception. If real memory is used it will be triggered at 95% of JVM heap, otherwise at 70% due to estimates. I guess both are configurable (?). Do you think we need to clarify anything here? Just mention the settings maybe?

radu-gheorghe replied:
Right, 95% is the current default for real memory, 70% for "sum-of-all-circuit-breakers". I think we can safely assume people have real memory enabled, I didn't really see clusters without real memory. And yes, it's configurable.

As for clarifications, I think it's important to say that the parent CB triggered implies that the memory usage right now is very high.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+Suppose an Elasticsearch cluster experiences a sudden increase in query load or indexing throughput, leading to resource contention and triggering the parent circuit breaker. When this happens, the alert rule checks for instances of the parent circuit breaker being tripped over the last 5 minutes. The alert is triggered as soon as the circuit breaker is tripped at least once within the specified  ...
+
+#### Actions to take
+
```
More heap, nodes, too?

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+This alert rule continuously monitors the presence of unassigned shards in an Elasticsearch cluster, detecting anomalies in the number of unassigned shards over time. When anomalies are detected, it triggers a warning (WARN priority). The minimum delay between consecutive alerts triggered by this alert rule is set to 10 minutes.
+
+Suppose an Elasticsearch cluster typically maintains a low number of unassigned shards, but due to issues such as node failures or disk space constraints, the number of unassigned shards suddenly increases. When this happens, the alert rule checks for anomalies in the number of unassigned shards over the last 30 minutes. Upon detecting the anomaly, the alert rule triggers a warning.
```
Normally, you shouldn't have unassigned shards. But OK, if you have some, at least the number shouldn't change :)

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+This alert rule continuously monitors thread pool rejections in an Elasticsearch environment, detecting anomalies in the rate at which thread pool requests are rejected. When anomalies are detected, it triggers a warning (WARN priority). The minimum delay between consecutive alerts triggered by this alert rule is set to 10 minutes.
+
+Suppose an Elasticsearch cluster experiences a sudden increase in thread pool rejections, potentially due to resource limitations or high query loads. When this happens, the alert rule checks for anomalies in thread pool rejections over the last 90 minutes. Upon detecting the anomaly, the alert rule triggers a warning.
```
A healthy cluster should have none, but maybe this is normal e.g. you're ingesting at the limit of what the cluster can handle. In this case, we should have at least a relatively constant number

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+#### Actions to take
+
+- Review system metrics for the Elasticsearch cluster, including CPU, memory, and disk usage, for any resource constraints that may be contributing to thread pool rejections
+- Analyze query patterns for any inefficient or resource-intensive queries. Optimize queries to reduce the load on the cluster
```
This only applies for the search thread pool

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+- Review system metrics for the Elasticsearch cluster, including CPU, memory, and disk usage, for any resource constraints that may be contributing to thread pool rejections
+- Analyze query patterns for any inefficient or resource-intensive queries. Optimize queries to reduce the load on the cluster
+- Review thread pool configurations and adjust settings such as thread counts, queue sizes, and timeouts to better accommodate the workload and prevent thread pool saturation
```
This is tricky advice, because we don't want to tell people to increase queues and just postpone the problem. But maybe we can tell people to check the calling apps and use fewer threads to talk to Elasticsearch. On the indexing side, exponential backoff is a good practice.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+This alert rule continuously monitors memory usage in an Elasticsearch environment and triggers a warning (WARN priority) when the used memory exceeds 80% of the total available memory. The minimum delay between consecutive alerts triggered by this alert rule is set to 10 minutes.
+
+Suppose the Elasticsearch cluster experiences a sudden increase in activity or data volume, leading to increased memory usage. When the used memory exceeds 80% of the total available memory, the alert rule checks memory usage over the last hour. Upon crossing the threshold, the alert rule triggers a warning.
```
Data volume shouldn't increase memory usage. Elasticsearch should use heap + a few other things. If there's more than 80% memory used => either heap is set too high or something else is using memory. Or some weird bug.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+#### Actions to take
+
+- Investigate the cause of increased memory usage in the Elasticsearch cluster, such as processes or activities contributing to the high memory consumption
+- Review and optimize the configuration settings of the Elasticsearch cluster, including heap size allocation and cache settings
```
Just heap size. Cache doesn't matter.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+The minimum delay between consecutive alerts triggered by this alert rule is set to 10 minutes.
+
+Suppose there is some activity detected in the swap usage on a node in the Elasticsearch cluster. Despite the relatively small amount of swap activity, the alert rule triggers a warning to prevent any big (and potentially unacceptable) slowdowns in Elasticsearch caused by accessing swap memory.
```
Yeah, you don't want to swap heap memory -> things can get very slow.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+Suppose there is some activity detected in the swap usage on a node in the Elasticsearch cluster. Despite the relatively small amount of swap activity, the alert rule triggers a warning to prevent any big (and potentially unacceptable) slowdowns in Elasticsearch caused by accessing swap memory.
+
+#### Action to take
```
Just disable swap, that's the only thing that can be done. Well, flavors: swapoff or vm.swappiness=1

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+Suppose an Elasticsearch cluster typically operates with a healthy percentage of open files, but due to increased usage or resource limitations, the percentage of open files exceeds 85%. When this happens, the alert rule checks for instances where the percentage of open files exceeds 85% within the last 10 minutes and triggers a warning.
+
+#### Actions to take
```
Make sure the limit of open files is correct (65K, I think, is recommended). If you get anywhere close, you have a big problem: weird merge policy, tons of very small shards maybe... never saw that situation. Changing directory (e.g. to hybridfs) should help, as well as increasing compound format percentage, force-merging old data.... but again, 99% it's a misconiguration.

radu-gheorghe commented on docs/integration/elasticsearch-integration.md:
```diff
@@ -52,6 +52,136 @@ Last, but certainly not least, you may want to get an alert if a node leaves the
+
+#### Actions to take
+
+-  Review system metrics such as CPU, memory, and disk usage for any source of increased load on the Elasticsearch cluster
```
not memory or disk, but everything related to load: thread pools, indexing and search operations, heap usage...

radu-gheorghe commented on docs/integration/kafka.md:
```diff
@@ -161,6 +161,104 @@ The Sematext Kafka monitoring agent collects the following metrics.
+#### Actions to take
+
+- Review Kafka consumer client configurations, message processing logic and system metrics to find the underlying cause of the increased lag
+- If the increased lag is due to limited resources, consider scaling up Kafka consumer client resources such as CPU, memory, or network bandwidth
```
Or increase the number of partitions and consumers?

radu-gheorghe commented on docs/integration/kafka.md:
```diff
@@ -161,6 +161,104 @@ The Sematext Kafka monitoring agent collects the following metrics.
+
+- Check the status of Kafka brokers for any broker failures or issues that may be causing partitions to become under-replicated
+- Review the replication configurations for affected topics and partitions and double check that replication factors are set correctly
+- AAddress any network issues or connectivity problems between Kafka brokers that may be impacting the replication of partitions
```
s/AAddress/Address/

radu-gheorghe commented on docs/integration/kafka.md:
```diff
@@ -161,6 +161,104 @@ The Sematext Kafka monitoring agent collects the following metrics.
+- Check the status of Kafka brokers for any broker failures or issues that may be causing partitions to become offline
+- Review the replication status and configuration of affected partitions to understand why they became offline. Determine if there are any replication issues or failures
+- If offline partitions are due to broker failures, try to recover or replace the failed brokers and make sure that partitions are redistributed and replicated properly
+- If offline partitions are due to disk failures or storage issues, address the underlying disk failures and mae sure that Kafka data directories are accessible and properly configured
```
s/mae/make/

radu-gheorghe commented on docs/integration/opensearch-integration.md:
```diff
@@ -40,6 +40,137 @@ OpenSearch runs within a Java Virtual Machine (JVM) and [monitoring JVM](https:/
 Finally, high disk reads and writes can indicate a poorly tuned system. Since accessing the disk is an expensive process in terms of time, a well-tuned system should reduce disk I/O wherever possible.
 
+
+## OpenSearch Default Alerts
```
I assume those are the same as Elasticsearch for the moment.

cpipilas replied:
Yes, that's  right

radu-gheorghe commented on docs/integration/solr-integration.md:
```diff
@@ -97,6 +97,24 @@ The Sematext Solr monitoring agent collects the following metrics.
+#### Actions to take
+
+- Review the warmup process of the Solr cache for any bottlenecks or configuration issues contributing to the extended warmup time
+- Optimize the warmup process by optimizing queries or adjusting cache configuration settings to reduce warmup time
```
You'd have to adjust autoWarmCount for some caches or the newSearcher queries. But maybe it's also that people do soft commits too often (either autoSoftCommit or from the application).

radu-gheorghe reviewed (changes requested):
I've done a full pass and left some comments, though I see you resolved most of them. Let me know if you need another pass :)

cpipilas commented:
> I've done a full pass and left some comments, though I see you resolved most of them. Let me know if you need another pass :)

Thank you, thank you! I think I've addressed all the comments, I've just requested a review to make sure there are no more comments or clarifications.

radu-gheorghe commented:
LGTM, thanks for adjusting!
